internal/controllers/gateway/route_utils.go

```diff
@@ -280,22 +280,44 @@ func routeHostnamesIntersectsWithListenerHostname[T types.RouteT](route T, liste
 	}
 }
 
-func routeTypeMatchesListenerType[T types.RouteT](route T, listenerProtocol ProtocolType) bool {
+func routeTypeMatchesListenerType[T types.RouteT](route T, listener Listener) bool {
 	switch (any)(route).(type) {
 	case *gatewayv1beta1.HTTPRoute:
-		if !(listenerProtocol == HTTPProtocolType || listenerProtocol == HTTPSProtocolType) {
+		// HTTPRoutes support Terminate only
+		// Note: this is a guess we are doing as the upstream documentation is unclear at the moment.
+		// see https://github.com/kubernetes-sigs/gateway-api/issues/1474
+		if !(listener.Protocol == HTTPProtocolType || listener.Protocol == HTTPSProtocolType) {
+			return false
+		}
+		if listener.TLS != nil && *listener.TLS.Mode != gatewayv1beta1.TLSModeTerminate {
 			return false
 		}
 	case *gatewayv1alpha2.TCPRoute:
-		if listenerProtocol != TCPProtocolType {
+		if listener.Protocol != TCPProtocolType {
+			return false
+		}
+		// TCPRoutes support Terminate only
+		// Note: this is a guess we are doing as the upstream documentation is unclear at the moment.
+		// see https://github.com/kubernetes-sigs/gateway-api/issues/1474
+		if listener.TLS != nil && *listener.TLS.Mode != gatewayv1beta1.TLSModeTerminate {
 			return false
 		}
 	case *gatewayv1alpha2.UDPRoute:
-		if listenerProtocol != UDPProtocolType {
+		if listener.Protocol != UDPProtocolType {
+			return false
+		}
+		// TLS should not be set in UDP listeners
+		if listener.TLS != nil {
 			return false
 		}
 	case *gatewayv1alpha2.TLSRoute:
-		if listenerProtocol != TLSProtocolType {
+		if listener.Protocol != TLSProtocolType {
+			return false
+		}
+		// TLSRoutes currently support Passthrough only
+		// Note: this is a guess we are doing as the upstream documentation is unclear at the moment.
+		// see https://github.com/kubernetes-sigs/gateway-api/issues/1474
+		if listener.TLS != nil && *listener.TLS.Mode != gatewayv1beta1.TLSModePassthrough {
```

Comment:
I think it is better to take a more compatible option if the upstream specification is inconsistent or not clear enough. So I think the better solution here is to support both `Passthrough` and `Terminate` for both `TCPRoute` and `TLSRoute`.  
Also, the upstream doc https://gateway-api.sigs.k8s.io/guides/tls/#clientserver-and-tls providing a stricter compatibility matrix did not say the matrix includes **ALL** compatible situations, so it is better to follow https://gateway-api.sigs.k8s.io/concepts/api-overview/#route-summary-table.

Reply:
The issue ACs are about this one:
> - Attaching a HTTPRoute or TCPRoute to a Gateway that supports TLS but it wants to use passthrough, should be resolved to an error.
> - Attaching a TLSRoute to a TLS listener that specifies terminate results in an error condition.

We discussed this in a meeting a while ago and agreed on scoping down the tls mode to the ACs described in the issue. When the upstream issue is solved, we'll change it accordingly.